xlcalculator/xlfunctions/financial.py

```diff
@@ -193,6 +220,66 @@ def VDB(
     return result
 
 
+@xl.register()
+@xl.validate_args
+def XIRR(
+        values: func_xltypes.XlArray,
+        dates: func_xltypes.XlArray,
+        guess: func_xltypes.XlNumber
+) -> func_xltypes.XlNumber:
+    """Returns the internal rate of return for a schedule of cash flows that
+        is not necessarily periodic.
+
+    https://support.microsoft.com/en-us/office/
+        xirr-function-de1242ec-6477-445b-b11b-a303ad9adc9d
+
+    Algorithm found on stackoverflow:
+    https://stackoverflow.com/questions/63797804/
+        python-irr-function-giving-different-result-than-excel-xirr
+
+    From MS, Newton's method is used to optimize:
+    https://docs.microsoft.com/en-us/office/troubleshoot/excel/
+        algorithm-of-xirr-funcation
+    """
+    def xnpv(rate, values, dates):
+        if rate <= -1.0:
+            return float('inf')
+
+        min_date = min(dates)
+
+        return sum([value / (1 + rate)**((date - min_date).days / 365)
+                   for value, date in zip(values, dates)])
+
+    def xirr(values, dates, guess):
+        try:
+            return newton(lambda r: xnpv(r, values, dates), guess, maxiter=100)
+
+        except (RuntimeError, FloatingPointError):
+            raise xlerrors.NumExcelError('XIRR did not converge')
```

Comment:
I would make these private functions of the module. It makes this function more readable and allows for smaller unit tests.

Reply:
xnpv is also duplicated where it probably doesn't need to be.